File: week6/credit/credit.py
import cs50
from cs50 import get_int
from math import floor
import sys 


def main():
    
    d1 = 0
    total = 0
    odds = 0
    evens = 0
    
    # Card number input
    number = get_int("Number: ")
    sum_d = len(str(number))
    
    # Read number as string to access each digit 
    num_s = str(number)
    
    # Check number of digits
    if sum_d != 13 and sum_d != 15 and sum_d != 16:
        print("INVALID")
        sys.exit()
    
    while number > 0:
        # Initialize digits and start from the end
        d1 = number % 10
        
        # Sum even digits 
        if total % 2 == 0:
            evens += d1
        
        # Sum multiply odd digits by 2, separate two digits and sum each
        else:
            double = d1 * 2
            odds += (double // 10) + (double % 10)
        
        # Move to the next digit from the end
        number //= 10
        total += 1
    
    # Sum adds and evens and if checksum is even 
    checksum = (odds + evens)
    if checksum % 2 == 0:
        # Access first two digit, check and match to each card type
        if int(num_s[0]) == 3 and int(num_s[1]) == 4 or int(num_s[1]) == 7:
            print("AMEX")
            
        elif int(num_s[0]) == 5 and int(num_s[1]) == 1 or int(num_s[1]) == 2 or int(num_s[1]) == 3 or int(num_s[1]) == 4 or int(num_s[1]) == 5:
            print("MASTERCARD")
            
        elif int(num_s[0]) == 4:
            print("VISA")
            
        # Invalid if no match    
        else:
            ("INVALID")


if __name__ == "__main__":
    main()            

File: week1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    //Get input as card number 
    long card_number = get_long("Number: ");
    
    //Count length of the number
    int i = 0;
    long num = card_number;
    while (num > 0)
    {
        num = num / 10;
        i++;
    }
    
    //Check the length to be 13,15 or 16
    if (i != 13 && i != 15 && i != 16)
    {
        printf("INVALID\n");
        return 0;
    }
    
    //Checksum
    int sum = 0;
    int sum2 = 0;
    long n = card_number; 
    int total = 0; 
    int mod1;
    int mod2; 
    int dig1;
    int dig2;
    
    do
    {
        //Add last number to sum
        mod1 = n % 10;
        n = n / 10;
        sum = sum + mod1;
        
        //second last digit 
        mod2 = n % 10;
        n = n / 10;
        
        //multiply by two and sum 
        mod2 = mod2 * 2;
        dig1 = mod2 % 10;
        dig2 = mod2 / 10;
        sum2 = sum2 + dig1 + dig2;
    }
    while (n > 0);
    total = sum + sum2; 
    
    if (total % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    
    //Find starting number 
    long star_num = card_number;
    do  
    {
        star_num = star_num / 10;
    }
    while (star_num > 100);
    
    if ((star_num / 10 == 5) && (star_num % 10 > 0 && star_num % 10 < 6))
    {
        printf("MASTERCARD\n");
    }
    else if ((star_num / 10 == 3) && (star_num % 10 == 4 || star_num % 10 == 7))
    {
        printf("AMEX\n");
    }
    else if (star_num / 10 == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}